Feature/settings organs

oleg68 opened the pull request:
Resolves: #700
Resolves: #886
Resolves: #950
Resolves: #1049
Resolves: #1119

This PR summarises a quite large amount of work on the Organs tab of the Settings dialog.

- The package list is got rid. The packages are maintained automatically with the corresponded organs #950
- Several organs may be selected a time. It allows to do same actions with several organs at once
- Additional information is shown including the `Organ hash` that allows to identify GrandOrgue files in the Cache and Data folders (#700)
- Capability to delete cache and saved presets for organs
- Capability to change the filesystem path of organs after the files were moved

larspalo commented:
There are a few small spelling errors and lost words in the strings that are easily fixed, but I think the UI experience (especially the behavior of the "Delete cache" and "Delete preset" buttons) needs some refinement before this PR is ready.

1. The buttons should only be enabled if the currently selected organ actually has a registered cache (for the "Delete cache") or preset (for the "Delete preset"). 
2. If no cache/preset is present, the corresponding button should remain/become inactive to denote that nothing can be done. 
3. When selecting another organ a new logic check should be made to see what buttons should be active (or not).

oleg68 commented:
> There are a few small spelling errors and lost words in the strings that are easily fixed, but I think the UI experience (especially the behavior of the "Delete cache" and "Delete preset" buttons) needs some refinement before this PR is ready.
> 
> 1. The buttons should only be enabled if the currently selected organ actually has a registered cache (for the "Delete cache") or preset (for the "Delete preset").
> 2. If no cache/preset is present, the corresponding button should remain/become inactive to denote that nothing can be done.
> 3. When selecting another organ a new logic check should be made to see what buttons should be active (or not).

1. I afraid that file check operations are quite expensive for navigations across the organ list, so I'd prefer to do this checking only after pressing the buttons.
2. Enabling/disabling logic may confuse the user when multiple organs are selected at the same time. Now the cache/presets are deleted for all selected organs if they are present

larspalo reviewed (changes requested):
CHANGELOG.md - both topics reference the same issue that's not really directly related to them, perhaps more clarification is needed?

help/grandorgue.xml:
line 56, 57, 66 comments?
line 137 _("Package name:")
line 151 _("Path in the package:")
line 165 _("Package hash:")
line 179 _("Package info:")
line 206 _("Delete &cache(s)")
line 207 _("Delete &preset(s)")
Logic of buttons to actually check if they are useful if enabled - not only if an organ line is selected.
line 694 "Do you want to delete all presets...

line 4974 punctuation mark lost?
line 4977 - this needs some clarification. Organs can be registered in multiple ways. Known organ packages automatically registers when starting GO.
line 4983 - change to "at least one line is"
line 4990 entry/entries
line 4994 entry/entries
line 4998 entry/entries
line 5015 useful
line 5017 no longer valid.
line 5025 punctuation mark?
line 5032 punctuation mark?
line 5038 clarification needed that also any orphaned entries found will be deleted at the same time.
line 5040 package

GOSettingsOrgans.cpp

larspalo commented:
> * I afraid that file check operations are quite expensive for navigations across the organ list, so I'd prefer to do this checking only after pressing the buttons.
> 
> * Enabling/disabling logic may confuse the user when multiple organs are selected at the same time. Now the cache/presets are deleted for all selected organs if they are present

Still it's bad UI design and I would happily sacrifice the multiple organs selection to fix it.

larspalo commented:
Seems like all my correction suggestions for the GOSettingsOrgans.cpp was lost...

oleg68 commented:
> CHANGELOG.md - both topics reference the same issue that's not really directly related to them, perhaps more clarification is needed?
I changed the issue reference to #1119. I separated deletion of subobjects to a separate entry because it is change of behavior that is not evident.

> help/grandorgue.xml: line 4974 punctuation mark lost? 
fixed

> line 4977 - this needs some clarification. Organs can be registered in multiple ways. Known organ packages automatically registers when starting GO. 
I described these ways

> line 4983 - change to "at least one line is" 
fixed

> line 4990 entry/entries line; 4994 entry/entries; line 4998 entry/entries 
fixed

> line 5015 useful line 5017 no longer valid. 
fixed

> line 5025 punctuation mark? line 5032 punctuation mark? 
fixed

> line 5038 clarification needed that also any orphaned entries found will be deleted at the same time.
Now it is not implemented in this way. The button only deletes files of the selected organs only.

May be to make a separate button for deleting orphaned files are not belonging to any registered organs?

>  line 5040 package
fixed

oleg68 commented:
> GOSettingsOrgans.cpp
> line 56, 57, 66 comments?

Deleted

> line 137 _("Package name:")

fixed

> line 151 _("Path in the package:")

fixed

> line 165 _("Package hash:")

fixed

> line 179 _("Package info:")

fixed

> line 206 _("Delete &cache(s)")

fixed

> line 207 _("Delete &preset(s)")

fixed

> Logic of buttons to actually check if they are useful if enabled - not only if an organ line is selected.

I'd like to avoid any filesystem-level checks when rendering UI

> line 694 "Do you want to delete all presets...

fixed

larspalo commented:
> Now it is not implemented in this way.

When I tested, no missing or incomplete organs were listed, but when I deleted one organ all the missing ones did appear in a comma separated list after my selected and were removed. From my quick födelsedagen review this is actually what happens, even if you didn't intend it. I have no problem with that but it should be documented.

larspalo commented:
> I'd like to avoid any filesystem-level checks when rendering UI

It's a sure sign of bad programming when the user can click a button for an action that has no action. The checks are made anyway (somewhere down the line) and it would not be a big problem solving it so that one can decide if the buttons should be active for the selected organ or any one in the selected range.

oleg68 commented:
> > Now it is not implemented in this way.
> 
> When I tested, no missing or incomplete organs were listed, but when I deleted one organ all the missing ones did appear in a comma separated list after my selected and were removed. From my quick code review this is actually what happens, even if you didn't intend it. I have no problem with that but it should be documented.

The packages that are not longer referenced by existing organs (orphaned) are deleted. But any orphaned cache and presets are not deleted unless their organs were deleted in this dialog session.

oleg68 commented:
> > I'd like to avoid any filesystem-level checks when rendering UI
> 
> It's a sure sign of bad programming when the user can click a button for an action that has no action. The checks are made anyway (somewhere down the line) and it would not be a big problem solving it so that one can decide if the buttons should be active for the selected organ or any one in the selected range.

Searching the files for existence by wildcards may take a long time so it shouldn't be done untill the user presses the button.

larspalo commented:
> Searching the files for existence by wildcards may take a long time so it shouldn't be done untill the user presses the button.

No, the buttons should not be possible to press at all unless that action indeed can be taken. If you're concerned about performance (which I highly doubt we should be, since it's a dialog not supposed to be used at the same time as one is playing the organ) there are ways to deal with that.

oleg68 commented:
> > But any orphaned cache and presets are not deleted unless their organs were deleted in this dialog session.
> 
> So, are you're saying we might have an issue with the missing or incomplete organs, that are not listed but might have caches and presets created for them?

Yes, I am. For example, if the organ was deleted with a previous version of GrandOrgue, where the files were not deleted automatically.

May be trying to delete the orphaned files every time the user closes the dialog with OK?

oleg68 commented:
> > Searching the files for existence by wildcards may take a long time so it shouldn't be done untill the user presses the button.
> 
> No, the buttons should not be possible to press at all unless that action indeed can be taken. If you're concerned about performance (which I highly doubt we should be, since it's a dialog not supposed to be used at the same time as one is playing the organ) there are ways to deal with that.

I usually do not sacrify performance in favor of UI. If opening the dialog or navigating across organs slowed down for a few seconds it would be much worse than unnecessary enabling the buttons.

So I have two reasons against this feature
1. Performance
2. Confusing with multi-selection

larspalo commented:
Also it's a bit confusing that additional organs in a package is not really reported as belonging to that package. Any idea about that?

oleg68 commented:
> Also it's a bit confusing that additional organs in a package is not really reported as belonging to that package. Any idea about that?

How can I reproduce this?

oleg68 commented:
> There are just a few more small things I'm going to suggest to change in the text,

I ask you to use `Suggest change` with the Github UI instead of enumerating them in a comment.

larspalo commented:
> > Also it's a bit confusing that additional organs in a package is not really reported as belonging to that package. Any idea about that?
> 
> How can I reproduce this?

This also seems to be related to if the organ package was registered with an earlier version of GO. If I remove it an re-registers it, then it has the correct information.

However, I did notice that the package information was not updated when I removed the package. The package was indeed removed from the list but the information in the textfields was not updated when no file now was "selected". Maybe the selection after the delete should be enforced to another in the list, or at least clear all the fields below.

Nor is the additional organs from a package removed when I remove the main organ of the package, but if I then remove the additional organ it asks me if I want to remove the main organ too - which I already had done. This could perhaps be improved upon.

larspalo commented on help/grandorgue.xml:
```diff
@@ -4967,39 +4967,42 @@ MIDI settings work properly.
+	<para>
+	  Organs can be registered in multiple ways. When a user opens an .organ
+	  file first time, the organ is registered. When the user installs an
+	  organ packages (.orgue) file, all organs presenting in the package are
```
```suggestion
	  organ package (.orgue) file, all organs present in the package are
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -4967,39 +4967,42 @@ MIDI settings work properly.
+	<para>This tab manages the <emphasis>known organs list</emphasis>.</para>
 	<para>All organs successfully loaded via the <link linkend="file_open">File/Open</link> menu item are automatically registered in this list.</para>
 	<para>
+	  An additional information about each organ is shown on the bottom part of this tab
```
```suggestion
	  Additional information about each organ is shown on the bottom part of this tab
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -4967,39 +4967,42 @@ MIDI settings work properly.
+	  file first time, the organ is registered. When the user installs an
+	  organ packages (.orgue) file, all organs presenting in the package are
+	  registered. Also .orgue packages placed in the packages directory are
+	  read upon GrandOrgue startup and all their organs are also registered
```
```suggestion
	  read upon GrandOrgue startup and all their organs are also registered.
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -5009,32 +5012,38 @@ MIDI settings work properly.
+	    This button opens a file choosing dialog for specifying the new
+	    path to the file the organ was opened from. It is useful after
+	    moving the organ files to another directory, and the orignal path is
+	    no longer valid
```
```suggestion
	    no longer valid.
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -5009,32 +5012,38 @@ MIDI settings work properly.
+	<sect3>
+	  <title>Delete Cache</title>
+	  <para>
+	    This button removes all cache files for this organs.
```
```suggestion
	    This button removes all cache files for the selected organ(s).
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -5009,32 +5012,38 @@ MIDI settings work properly.
+	<sect3>
+	  <title>Delete Presets</title>
+	  <para>
+	    This button removes all setting files for this organs that were
```
```suggestion
	    This button removes all setting files for the selected organ(s) that were
```

larspalo commented on help/grandorgue.xml:
```diff
@@ -5009,32 +5012,38 @@ MIDI settings work properly.
+	  <para>
+	    This button removes all setting files for this organs that were
+	    saved with the Save menu entry. All the user settings for the organ
+	    will be lost
```
```suggestion
	    will be lost.
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
-  m_Organs->InsertColumn(4, _("Organ package"));
-  m_Organs->InsertColumn(5, _("ODF Path"));
-  boxOrgans->Add(m_Organs, 1, wxEXPAND | wxALL, 5);
+    //    wxSize(400, 200),
```
```suggestion

```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  gbSizer->Add(m_Recording, wxGBPosition(2, 1), wxDefaultSpan, wxEXPAND);
+
+  gbSizer->Add(
+    new wxStaticText(this, wxID_ANY, _("Organ Hash:")),
```
```suggestion
    new wxStaticText(this, wxID_ANY, _("Organ hash:")),
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  gbSizer->Add(m_OrganHash, wxGBPosition(3, 1), wxDefaultSpan, wxEXPAND);
+
+  gbSizer->Add(
+    new wxStaticText(this, wxID_ANY, _("Package Id:")),
```
```suggestion
    new wxStaticText(this, wxID_ANY, _("Package id:")),
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  gbSizer->Add(m_PackageId, wxGBPosition(4, 1), wxDefaultSpan, wxEXPAND);
+
+  gbSizer->Add(
+    new wxStaticText(this, wxID_ANY, _("Package Name:")),
```
```suggestion
    new wxStaticText(this, wxID_ANY, _("Package name:")),
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  gbSizer->Add(m_PackageName, wxGBPosition(1, 3), wxDefaultSpan, wxEXPAND);
+
+  gbSizer->Add(
+    new wxStaticText(this, wxID_ANY, _("Path in the Package:")),
```
```suggestion
    new wxStaticText(this, wxID_ANY, _("Path in package:")),
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+    std::set<wxString> packagePathsUsed;
+
+    for (long l = m_Organs->GetItemCount(), i = 0; i < l; i++)
+      if (selectedItemSet.find(i) == selectedItemSet.end()) { // not selected
```
```suggestion
      if (selectedItemSet.find(i) == selectedItemSet.end()) {
        // not selected
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  GOOrgan *pOrgan = (GOOrgan *)m_Organs->GetItemData(index);
+  auto oldHashEntry = m_OldHashes.find(pOrgan);
+  const wxString oldHash = oldHashEntry == m_OldHashes.end()
+    ? pOrgan->GetOrganHash() // the first relocating of this organ
```
```suggestion
    ? pOrgan->GetOrganHash()
    // the first relocating of this organ
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  auto oldHashEntry = m_OldHashes.find(pOrgan);
+  const wxString oldHash = oldHashEntry == m_OldHashes.end()
+    ? pOrgan->GetOrganHash() // the first relocating of this organ
+    : oldHashEntry->second;  // this organ has already been relocated. Use the
```
```suggestion
    : oldHashEntry->second;
    // this organ has already been relocated. Use the
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+  const wxString oldHash = oldHashEntry == m_OldHashes.end()
+    ? pOrgan->GetOrganHash() // the first relocating of this organ
+    : oldHashEntry->second;  // this organ has already been relocated. Use the
+                             // original hash
```
```suggestion
    // original hash
```

larspalo commented on src/grandorgue/dialogs/settings/GOSettingsOrgans.cpp:
```diff
@@ -7,324 +7,775 @@
+
+    if (!GetPkgByPath(origPackage->GetPath()))
+      delete origPackage;
+    origPackage = NULL; // for preventing deleting during clear()
```
```suggestion
    // for preventing double deleting during clear()
    origPackage = NULL;
```

oleg68 commented:
@larspalo I've fixed your CR notes.

There are a still disscussed question about enabling/disabling Delete cache(s)/preset(s) button. I thing this PR is more focused on the code, not on the concepts. So I invite you to #955. There are also some ideas from @paolo2504. Lets continue discussing there

larspalo commented:
I approve of all the changes so far, but as more improvements are coming (I guess, from the discussion, right?) I'll wait to approve this PR.

oleg68 commented:
@larspalo I've made all the changes I wanted for this PR. You can test the new version from https://github.com/oleg68/GrandOrgue-official/actions/runs/2682478835

larspalo commented:
@oleg68 It seems to be some issue with the latest action build for this PR in your repository.

oleg68 commented:
> It seems to be some issue with the latest action build for this PR in your repository.

Yes, it is. The issue is not related to the changes of this PR, so it will be resolved with a separate PR.

larspalo reviewed (approved):
As far as I've been able to test it, this PR works exactly as intended now. Great work! The only little small spot is the screenshot for the help that contain a few of the old spelling errors and some slightly different strings than the new fixed uses. I don't consider it to be a stopper though, so I approve of this PR, it's certainly a great step forward in the internal file management of GO!

oleg68 commented:
@larspalo Now I'm working on [the last issue](https://github.com/GrandOrgue/grandorgue/discussions/955#discussioncomment-3174399) mentioned by @paolo2504

rousseldenis reviewed (approved):
Tested. I like the improvements. Thanks!
